### PR TITLE
Fix some log messages about XML loading

### PR DESCRIPTION
**Pull request by rldhont**
Firstly fix `Object of class LibXMLError could not be converted to string` in OGCRequest.

Secondly add error log messages about QGIS Project XML loading

Then add error log messages about WMS GetCapabilities loading in WMTS classes.

* Funded by CRIGE-PACA

### CODE REVIEW

**Review by nboisteault — APPROVED** [review]
@rldhont it looks good to me but we have no project to replicate the issue and test your PR.

**Review by mdouchin — CHANGES_REQUESTED** [review]
I think we should add a method in a utility class to manage the building of the `$errormsg` to avoid to duplicate four times the code (with small changes). Probably could be a method of `lizmap/modules.lizmap/lib/Request/Proxy.php` or in another more appropriate class.

File: lizmap/modules/lizmap/classes/lizmapTiler.class.php [hunk 1]
```diff
@@ -80,15 +80,54 @@ public static function getTileCapabilities($project)
...
+                $errormsg = 'An error has been raised when loading WMS GetCapabilities: ServiceExceptionReport';
+                $errormsg .= '\n'.$wms;
+                \jLog::log($errormsg, 'error');
+
                 return null;
             }
 
+            $use_errors = libxml_use_internal_errors(true);
             $wms_xml = simplexml_load_string($wms);
+            if ($wms_xml === false) {
```
[the remark this answers is not in the document]
Comment (rldhont) [reply lizmap/modules/lizmap/classes/lizmapTiler.class.php:100]: Ok I'll do

[the remark this answers is not in the document]
Comment (rldhont) [reply lizmap/modules/lizmap/classes/lizmapTiler.class.php:99]: done

**Comment by rldhont** [comment]
@nboisteault @mdouchin can you re-review the code ?

**Review by mdouchin — APPROVED** [review]
👍 thanks

**Review by nboisteault — APPROVED** [review]
Could you add some unit tests?

**Review by laurentj — CHANGES_REQUESTED** [review]
These new methods are not related at all to Lizmap, and so should not be into the `lizmap` class. I would prefer to have them in the `tools` class or better, into a new `\Lizmap\App\XmlTools` class (into the lib directory of the lizmap module).

**Comment by rldhont** [comment]
@laurentj thanks for information

**Comment by rldhont** [comment]
@laurentj new class `\Lizmap\App\XmlTools` created with tests

@nboisteault  unit tests addded

File: lizmap/modules/lizmap/lib/App/XmlTools.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * app context informations inside a Jelix environment.
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/App/XmlTools.php:3]: wrong comment, and wrong copyright year ;-)

File: lizmap/modules/lizmap/lib/App/XmlTools.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ */
+
+namespace Lizmap\App;
+
+class XmlTools
+{
+    /*
+     * Interprets a string of XML into an object
+     *
+     * @param $xml_str a well-formed XML string
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/App/XmlTools.php:20]: `string` is missing. The good syntax is `@param string $xml_str  a well-formed XML string`

File: lizmap/modules/lizmap/lib/App/XmlTools.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return ($xml, $msg) Array containing $xml an object of class SimpleXMLElement with properties
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/App/XmlTools.php:22]: the good syntax here is : 

```
* @return array  Array containing a SimpleXMLElement object with properties containing the data held 
*                                  within the XML document, or false on failure ; and a string containing the error message.
```

File: lizmap/modules/lizmap/lib/App/XmlTools.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        $msg = '';
+        if (!$xml) {
+            $msg = self::xmlErrorMsg();
+        }
+
+        return array($xml, $msg);
+    }
+
+    /*
+     * Interprets an XML file into an object
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/App/XmlTools.php:39]: same issue in `@param` and `@return` syntax. See https://docs.phpdoc.org/latest/guide/guides/types.html

File: lizmap/modules/lizmap/lib/App/XmlTools.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        return array($xml, $msg);
+    }
+
+    /*
+     * Get XML error message
+     *
+     * Build an error message based on LibXMLError object
+     *
+     * @return $msg a string containing the error message.
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/App/XmlTools.php:64]: `@return string  the error message.`

File: lizmap/modules/lizmap/lib/Request/OGCRequest.php [hunk 7]
```diff
@@ -312,23 +312,26 @@ protected function process_getcapabilities()
         );
     }
 
+    /*
+     * Interprets a string of XML into an object
+     *
+     * @param $xmldata a well-formed XML string
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Request/OGCRequest.php:318]: `string` keyword is missing in both `@param`

**Review by laurentj — CHANGES_REQUESTED** [review]
It's ok to me, except DOC comments. It is important to respect the right syntax, to allow IDE to help on auto-completion and type checks, and so to improve the code quality. It helps also to generate a good API documentation (for the day when we will generate such documentations).

Probably we could improve the return type of `xmlFromString` and `xmlFromFile` by returning only one value (`@return SimpleXmlElement|string`). So, instead to do `if (!$xml)`, we could do `if(is_string($xml))` or `if(!is_object($xml))`.  With this returned value, there is not a unused variable (which exists here in both case, error or not), and it makes the signature of the method more readable.

I let you to decide to do the change or to keep as is, but at least  change the comments :-)

**Comment by Gustry** [comment]
> It is important to respect the right syntax, to allow IDE to help on auto-completion and type checks, and so to improve the code quality. It helps also to generate a good API documentation (for the day when we will generate such documentations).

Oh yes, big +1.
Can we revive #1755 in the mid term?

**Review by laurentj — APPROVED** [review]
LGTM. Thank you :-)

**Comment by rldhont** [comment]
Thanks to all

**Pull request merged**